aten/src/ATen/cpu/vec/vec512/vec512.h

```diff
@@ -0,0 +1,224 @@
+}
+
+#if defined(CPU_CAPABILITY_AVX512)
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ CAST (AVX2) ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+#define DEFINE_FLOAT_INT_CAST(int_t, float_t, float_ch)                   \
+template<>                                                                \
+inline  Vectorized<int_t> cast<int_t, float_t>(const Vectorized<float_t>& src) {  \
+  return _mm512_castp ## float_ch ## _si512(src);                         \
+}                                                                         \
+template<>                                                                \
+inline Vectorized<float_t> cast<float_t, int_t>(const Vectorized<int_t>& src) {   \
+  return _mm512_castsi512_p ## float_ch (src);                            \
+}
+
+DEFINE_FLOAT_INT_CAST(int64_t, double, d)
+DEFINE_FLOAT_INT_CAST(int32_t, double, d)
+DEFINE_FLOAT_INT_CAST(int16_t, double, d)
+DEFINE_FLOAT_INT_CAST(int64_t, float, s)
+DEFINE_FLOAT_INT_CAST(int32_t, float, s)
+DEFINE_FLOAT_INT_CAST(int16_t, float, s)
+
+#undef DEFINE_FLOAT_INT_CAST
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ GATHER ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template<int64_t scale = 1>
+std::enable_if_t<scale == 1 || scale == 2 || scale == 4 || scale == 8, Vectorized<double>>
+inline gather(const double* base_addr, const Vectorized<int64_t>& vindex) {
+  return _mm512_i64gather_pd(vindex, base_addr, scale);
+}
+
+template<int64_t scale = 1>
+std::enable_if_t<scale == 1 || scale == 2 || scale == 4 || scale == 8, Vectorized<float>>
+inline gather(const float* base_addr, const Vectorized<int32_t>& vindex) {
+  return _mm512_i32gather_ps(vindex, base_addr, scale);
+}
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ MASK GATHER ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template<int64_t scale = 1>
+std::enable_if_t<scale == 1 || scale == 2 || scale == 4 || scale == 8, Vectorized<double>>
+inline mask_gather(const Vectorized<double>& src, const double* base_addr,
+                   const Vectorized<int64_t>& vindex, const Vectorized<double>& mask) {
+  auto all_ones = _mm512_castsi512_pd(_mm512_set1_epi64(0xFFFFFFFFFFFFFFFF));
+  auto mask_ = _mm512_cmp_pd_mask(all_ones, mask.values, _CMP_EQ_OQ);
+  return _mm512_mask_i64gather_pd(src, mask_, vindex, base_addr, scale);
+}
+
+template<int64_t scale = 1>
+std::enable_if_t<scale == 1 || scale == 2 || scale == 4 || scale == 8, Vectorized<float>>
+inline mask_gather(const Vectorized<float>& src, const float* base_addr,
+                   const Vectorized<int32_t>& vindex, const Vectorized<float>& mask) {
+  auto all_ones = _mm512_castsi512_ps(_mm512_set1_epi32(0xFFFFFFFF));
+  auto mask_ = _mm512_cmp_ps_mask(all_ones, mask.values, _CMP_EQ_OQ);
+  return _mm512_mask_i32gather_ps(src, mask_, vindex, base_addr, scale);
+}
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ CONVERT ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+// Only works for inputs in the range: [-2^51, 2^51]
+// From: https://stackoverflow.com/a/41148578
+// We can revisit it if tests would nudge us to

```

Comment:
This looks out of date now, right? Because now we have the single instruction that does it.

Reply:
Yes, I'll revise all the comments tonight